src/bridge/bridge.c

```diff
@@ -173,79 +173,39 @@ send_init_command (CockpitTransport *transport,
   g_bytes_unref (bytes);
 }
 
-static void
-setup_dbus_daemon (gpointer addrfd)
-{
-  g_unsetenv ("G_DEBUG");
-  cockpit_unix_fd_close_all (3, GPOINTER_TO_INT (addrfd));
-}
-
-static GPid
+static GSubprocess *
 start_dbus_daemon (void)
 {
-  GError *error = NULL;
-  GString *address = NULL;
-  gchar *line;
-  gsize len;
-  gssize ret;
-  GPid pid = 0;
-  gchar *print_address = NULL;
-  int addrfd[2] = { -1, -1 };
-  GSpawnFlags flags;
-
-  gchar *dbus_argv[] = {
-      "dbus-daemon",
-      "--print-address=X",
-      "--session",
-      NULL
-  };
-
-  if (pipe (addrfd))
-    {
-      g_warning ("pipe failed to allocate fds: %m");
-      goto out;
-    }
-
-  print_address = g_strdup_printf ("--print-address=%d", addrfd[1]);
-  dbus_argv[1] = print_address;
+  g_autoptr(GError) error = NULL;
 
   /* The DBus daemon produces useless messages on stderr mixed in */
-  flags = G_SPAWN_LEAVE_DESCRIPTORS_OPEN | G_SPAWN_SEARCH_PATH |
-          G_SPAWN_STDERR_TO_DEV_NULL | G_SPAWN_STDOUT_TO_DEV_NULL;
-
-  g_spawn_async_with_pipes (NULL, dbus_argv, NULL, flags,
-                            setup_dbus_daemon, GINT_TO_POINTER (addrfd[1]),
-                            &pid, NULL, NULL, NULL, &error);
-
-  close (addrfd[1]);
+  g_autoptr(GSubprocessLauncher) dbus_daemon = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_STDOUT_PIPE |
+                                                                          G_SUBPROCESS_FLAGS_STDERR_SILENCE);
+  g_subprocess_launcher_unsetenv (dbus_daemon, "G_DEBUG");
+  GSubprocess *process = g_subprocess_launcher_spawn (dbus_daemon, &error, "dbus-daemon", "--print-address", "--session", NULL);
 
   if (error != NULL)
     {
       if (g_error_matches (error, G_SPAWN_ERROR, G_SPAWN_ERROR_NOENT))
-        g_debug ("couldn't start %s: %s", dbus_argv[0], error->message);
+        g_debug ("couldn't start dbus-daemon: %s", error->message);
       else
-        g_message ("couldn't start %s: %s", dbus_argv[0], error->message);
-      g_error_free (error);
-      pid = 0;
-      goto out;
+        g_message ("couldn't start dbus-daemon: %s", error->message);
+      return NULL;
     }
 
-  g_debug ("launched %s", dbus_argv[0]);
+  g_debug ("launched dbus-daemon: %s", g_subprocess_get_identifier (process));
 
-  address = g_string_new ("");
+  g_autoptr(GString) address = g_string_new ("");
   for (;;)
     {
-      len = address->len;
+      gsize len = address->len;
       g_string_set_size (address, len + 256);
-      ret = read (addrfd[0], address->str + len, 256);
+      gssize ret = g_input_stream_read (g_subprocess_get_stdout_pipe (process), address->str + len, 256, NULL, &error);
```

Comment:
`read_all()`

Reply:
Can't -- dbus-daemon does not close its stdout after writing the address, so read_all() will hang forever (I did try this). I'm afraid we have to read the full line. I was a bit disappointed that there's no `g_input_stream_read_line()` as an fgets() counterpart, that'd be really handy.

Reply:
`g_data_input_stream_read_line ()`

Reply:
Hah, thanks! So much nicer now. Pushed.